Revise README (#184)
XRPLF/xrpl-py: README.md

    "PreviousTxnLgrSeq": 16007556,
    "Sequence": 16007556,
    "index": "EEE14A3994B4067813CEF0E6B94F1FCD2B66204DEF1B8DB6E3E6909770132F7B"
}
```

[![Downloads](https://pepy.tech/badge/xrpl-py/month)](https://pepy.tech/project/xrpl-py/month)
[![Contributors](https://img.shields.io/github/contributors/xpring-eng/xrpl-py.svg)](https://github.com/xpring-eng/xrpl-py/graphs/contributors)

## Installation and supported versions

The `xrpl-py` library is available on [PyPI](https://pypi.org/). Install with `pip`:

```
pip3 install xrpl-py
```

The library supports [Python 3.7](https://www.python.org/downloads/) and later.

[![Supported Versions](https://img.shields.io/pypi/pyversions/xrpl-py.svg)](https://pypi.org/project/xrpl-py)

## Features

Use `xrpl-py` to build Python applications that leverage the [XRP Ledger](https://xrpl.org/). The library helps with all aspects of interacting with the XRP Ledger, including:

* Key and wallet management
* Serialization
* Transaction Signing

`xrpl-py` also provides:

* A network client — See [`xrpl.clients`](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.clients.html) for more information.

> * Methods for insepcting accounts — See [XRPL Account Methods](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.account.html) for more information.

Reviewer: ```suggestion
* Methods for inspecting accounts — See [XRPL Account Methods](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.account.html) for more information.
```

* Codecs for encoding and decoding addresses and other objects — See [Core Codecs](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.core.html) for more information.

## Usage

The following sections describe some of the most commonly used modules in the `xrpl-py` library and provide sample code.

> For complete reference documentation, see the [`xrpl-py` docs](https://xrpl-py.readthedocs.io/en/latest/index.html).

Reviewer: Note: the docs might be changed to `/stable` instead of `/latest`

> For complete reference documentation, see the [`xrpl-py` docs](https://xrpl-py.readthedocs.io/en/latest/index.html).

Author: I'll leave it as `/latest` for now but thanks for the heads up! Easy enough to change when the url does change.

### Network client

Use the `xrpl.clients` library to create a network client for connecting to the XRP Ledger.

```py
from xrpl.clients.json_rpc_client import JsonRpcClient
JSON_RPC_URL = "https://s.altnet.rippletest.net:51234/"
JSON_RPC_CLIENT = JsonRpcClient(JSON_RPC_URL)
```

### Manage keys and wallets

#### `xrpl.wallet`

Use the [`xrpl.wallet`](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.wallet.html) module to create a wallet from a given seed or or via a [Testnet faucet](https://xrpl.org/xrp-testnet-faucet.html).

> To create a wallet from a seed (in this case, the value genrated in [`xrpl.keypairs`](#xrpl-keypairs)):

Reviewer: ```suggestion
To create a wallet from a seed (in this case, the value generated in [`xrpl.keypairs`](#xrpl-keypairs)):
```

* [`safe_sign_transaction`](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.transaction.html#xrpl.transaction.safe_sign_transaction) — Signs a transaction locally. This method **does  not** submit the transaction to the XRP Ledger.

* [`send_reliable_submission`](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.transaction.html#xrpl.transaction.send_reliable_submission) — An implementation of the [reliable transaction submission guidelines](https://xrpl.org/reliable-transaction-submission.html#reliable-transaction-submission), this method signs a transaction locally, submits the transaction to the XRP Ledger, and then verifies that it has been included in a validated ledger (or has failed to do so). Use this method to submit transactions for production purposes.

```py
from xrpl.models.transactions import Payment
from xrpl.models.transactions.transaction import Transaction
from xrpl.transaction import send_reliable_submission
from xrpl.ledger import get_fee

FEE = get_fee(JSON_RPC_CLIENT)

# prepare the transaction
my_tx_payment = Payment(
    account=TESTNET_CLASSIC_ACCOUNT,

>     amount=_AMOUNT,

Reviewer: Should probably also note that this is in drops, not XRP

    destination=TESTNET_DESTINATION_ACCOUNT,
    signing_pub_key=TESTNET_WALLET.pub_key,

>     last_ledger_sequence=TESTNET_WALLET.next_sequence_num + 10,

Author: Updated but I got this error when I tested it out:

```
>>> current_validated_ledger = get_latest_validated_ledger_sequence(client)
>>> print(test_wallet)
seed: shA5izLnSNFxNwGMV1ar6WJnnsNbo
pub_key: 029234B7F15318156A762E5411E6E3FE5A170D49FFC50286C38D7E68AE16B7B412
priv_key: 0031DFF36F8F22287BAD4137DA182635F3E0F0F510698E50A3039BB5CE170B941A
classic_address: rMPUKmzmDWEX1tQhzQ8oGFNfAEhnWNFwz

>>> my_tx_payment = Payment(
...     account="rMPUKmzmDWEX1tQhzQ8oGFNfAEhnWNFwz",
...     amount="2200000",
...     destination="rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe",
...     last_ledger_sequence=current_validated_ledger + 20,
...     sequence=test_wallet.next_sequence_num,
...     fee="10",
... )
>>> my_tx_payment_signed = safe_sign_transaction(my_tx_payment,test_wallet)
>>> tx_response = send_reliable_submission(my_tx_payment, test_wallet, client)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/rgy/Git/xrpl-py/xrpl/transaction/reliable_submission.py", line 87, in send_reliable_submission
    f"Transaction failed, {result_code}: {result_message}"
xrpl.transaction.exceptions.XRPLReliableSubmissionException: Transaction failed, tefMAX_LEDGER: Ledger sequence too high.
```

    sequence=TESTNET_WALLET.next_sequence_num,
    fee=FEE,
)
# sign the transaction
my_tx_payment_signed = safe_sign_transaction(my_tx_payment,TESTNET_WALLET)

# submit the transaction
tx_response = send_reliable_submission(my_tx_payment, TESTNET_WALLET, JSON_RPC_CLIENT_TESTNET)
```

### Encode addresses

Use the [`xrpl.core.addresscodec`](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.core.addresscodec.html) to encode and decode addresses into the "classic" and X-address formats.

```py
# convert classic address to x-address
>>> from xrpl.core import addresscodec

> >>> TESNET_XADDRESS = addresscodec.classic_address_to_xaddress(TESTNET_CLASSIC_ACCOUNT, False, True)

Reviewer 2: The second argument is supposed to be the tag (an integer from 0 to 2³² inclusive), not a boolean, if you provide it. Changing to keyword arguments makes this section much easier to understand, and using hardcoded values here will make this code snippet stand alone (again, for the benefit of people skimming):

```suggestion
>>> testnet_xaddress = addresscodec.classic_address_to_xaddress("ra9cjuPRq7tmR3Uin3y6fUPKLZ317JLvbN", tag=0, is_test_network=True)
```

>>> print(TESNET_XADDRESS)
T7CiH636SkvWbU17fKdttaHE97SXQRW1NfMESd67VDcoZTn
```

## Documentation

> For complete reference documentation for the library, see [`xrpl-py` docs](https://xrpl-py.readthedocs.io/en/latest/index.html).

Reviewer 2: Per my suggestion above, I suggest moving this up above the "Usage" section so it doesn't end up down here where the tl;dr people don't see it. (Though, those people won't exactly be using it anyway, will they? 😝)

from xrpl.ledger import get_latest_validated_ledger_sequence

current_validated_ledger = get_latest_validated_ledger_sequence(client)

# prepare the transaction
# the amount is expressed in drops, not XRP
# see https://xrpl.org/basic-data-types.html#specifying-currency-amounts
my_tx_payment = Payment(
    account="rMPUKmzmDWEX1tQhzQ8oGFNfAEhnWNFwz",
    amount="2200000",
    destination="rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe",
    last_ledger_sequence=current_validated_ledger + 20,
    sequence=test_wallet.next_sequence_num,
    fee="10",
)
# sign the transaction
my_tx_payment_signed = safe_sign_transaction(my_tx_payment,test_wallet)

# submit the transaction
tx_response = send_reliable_submission(my_tx_payment, test_wallet, client)
```

#### Get fee from the XRP Ledger

In most cases, you should specify `10` for the `fee` field unless you have a strong reason not to. But if you want to get the [current transaction cost](https://xrpl.org/transaction-cost.html#current-transaction-cost) from the XRP Ledger, you can use the `get_fee` function:

```py
from xrpl.ledger import get_fee

FEE = get_fee(client)
print(FEE)

# print output
10
```

> #### Auto-filled fields

Author: @mvadari added this per your comment. Could you take a quick look before I merge?

```py
from xrpl.models.transactions import Payment
from xrpl.transaction import send_reliable_submission
from xrpl.ledger import get_latest_validated_ledger_sequence

current_validated_ledger = get_latest_validated_ledger_sequence(client)

# prepare the transaction
# the amount is expressed in drops, not XRP
# see https://xrpl.org/basic-data-types.html#specifying-currency-amounts
my_tx_payment = Payment(
    account="rMPUKmzmDWEX1tQhzQ8oGFNfAEhnWNFwz",
    amount="2200000",
    destination="rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe",
    last_ledger_sequence=current_validated_ledger + 20,
    sequence=test_wallet.next_sequence_num,
    fee="10",
)
# sign the transaction
my_tx_payment_signed = safe_sign_transaction(my_tx_payment,test_wallet)

# submit the transaction
tx_response = send_reliable_submission(my_tx_payment, test_wallet, client)
```

#### Get fee from the XRP Ledger

> In most cases, you should specify `10` for the `fee` field unless you have a strong reason not to. But if you want to get the [current transaction cost](https://xrpl.org/transaction-cost.html#current-transaction-cost) from the XRP Ledger, you can use the `get_fee` function:

Reviewer: #197 also adds in autofilling capabilities for transactions

Reviewer 2: ```suggestion
In most cases, you can specify the minimum [transaction cost](https://xrpl.org/transaction-cost.html#current-transaction-cost) of `"10"` for the `fee` field. But if you want to get the current load-based transaction cost from the network, you can use the `get_fee` function:
```
